src/Bicep.Core/Analyzers/Linter/Rules/WhatIfShortCircuitingRule.cs

```diff
@@ -38,150 +42,689 @@ public override string FormatMessage(params object[] values)
+                        .Apply(ParameterReferenceCollector.Instance, null))
+                    {
+                        moduleModelFlags[parameterUsedInCondition] = ParameterUsageFlags.UsedInResourceCondition;
+                    }
+                }
+
+                if (extensibilityResource.ApiVersion is { } apiVersion)
+                {
+                    foreach (var parameterUsedInApiVersion in ExpressionParser.ParseLanguageExpression(apiVersion)
+                        .Apply(ParameterReferenceCollector.Instance, null))
+                    {
+                        moduleModelFlags[parameterUsedInApiVersion] = ParameterUsageFlags.UsedInResourceCondition;
+                    }
                 }
+
+                // Because extensibility resource identifiers are discovered during what-if by calling the
+                // provider, skip analysis of identifiers here.
             }
 
-            resource.Name?.Accept(sentinelVisitor);
-            resource.SubscriptionId?.Accept(sentinelVisitor);
-            resource.ResourceGroup?.Accept(sentinelVisitor);
-            resource.Condition?.Accept(sentinelVisitor);
-            resource.Scope?.Accept(sentinelVisitor);
-            resource.ApiVersion?.Accept(sentinelVisitor);
+            return moduleModelFlags;
         }
 
-        private static Template GetTemplate(SemanticModel model)
+        private static (
+            ImmutableArray<DeploymentResourceWithParsedExpressions> armResources,
+            ImmutableArray<ExtensibleResource> extensibilityResources
+        ) GetModuleTemplateAndNestedResources(
+            Template template,
+            ResourceScope targetScope,
+            IReadOnlyDictionary<string, ITemplateLanguageExpression> parameters)
         {
             try
             {
-                var textWriter = new StringWriter();
-                using var writer = new SourceAwareJsonTextWriter(textWriter)
-                {
-                    // don't close the textWriter when writer is disposed
-                    CloseOutput = false,
-                    Formatting = Formatting.Indented
-                };
-                var (template, _) = new TemplateWriter(model).GetTemplate(writer);
+                // A synchronous wait should be fine here since TemplateEngine.ExpandNestedDeployments will only await
+                // three things:
+                //  - Downloading linked parameters
+                //  - Downloading linked templates
+                //  - Fetching the default API version to use in a template that uses an API profile
+                // All three of these will synchronously fail with an exception because the calling code is not
+                // supplying either a contentLinkResolver or an apiProfileResolver, but it feels dangerous to put in
+                // a sync wait and hope the invoked code doesn't change what it's doing!
+#pragma warning disable VSTHRD002 // Avoid problematic synchronous waits
```

Comment:
Flagging this for scrutiny: a synchronous wait should be fine here, but I only know that from reviewing where `TemplateEngine.ExpandNestedDeployments` may wait. That's a point-in-time assessment that is not reflected in the method signature, so I feel a bit squeamish about suppressing a warning over it.

Two options to avoid the suppression:
1. We could introduce an async abstraction for linters. Maybe not a great option because this of how often they are invoked in the VS Code extension, but I would be other theoretical linters could benefit from an async option.
2. We could add a parallel `TemplateEngine.ExpandNestedDeploymentsSync`, which would be a fair amount of work.